# Tasca 5: Exploració de les dades

Descripció :

>Familiaritza't amb les tècniques d'exploració de les dades mitjantçant la estructura de dades, Dataframe amb la llibreria Pandas.

## Nivell 1

* Exercici 1 :
Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un pandas Dataframe. Explora les dades que conté, i queda’t únicament amb les columnes que consideris rellevants.

In [2]:
# Render our plots inline
%matplotlib inline

import pandas as pd
import matplotlib.pyplot as plt

#plt.show(block=True)
#pd.describe_option()
plt.style.use('ggplot')
#plt.style.use('default')
plt.rcParams['figure.figsize'] = (15, 5)

# Data Expo 2009 - Airline on-time performance

The data consists of flight arrival and departure details for all commercial flights within the USA, from October 1987 to April 2008.

# Reading data from CSVs

The read_csv function takes a file in csv format and converts it to a Pandas DataFrame:

In [3]:
# Reading data from CSV: with CSV files a single line is needed to load the data:
DelayedFlights_df = pd.read_csv('DelayedFlights.csv')
# let's see the DataFrame
DelayedFlights_df

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0         2211.0        2225            WN  ...     4.0     8.0          0   
1         1002.0        1000            WN  ...     5.0    10.0          0   
2          804.0         750            WN  ...     3.0    17.0          0   
3         1959.0        1925            WN  ...     3.0    10.0          0   
4         2121.0        2110            WN  ...     4.0    10.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   
1936756   1446.0        1437            DL  ...    13.0    13.0          0   
1936757   1413.0        1418            DL  ...     8.0    11.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                       N         0           NaN           NaN      NaN   
1                       N         0           NaN           NaN      NaN   
2                       N         0           NaN           NaN      NaN   
3                       N         0           2.0           0.0      0.0   
4                       N         0           NaN           NaN      NaN   
...                   ...       ...           ...           ...      ...   
1936753                 N         0           3.0           0.0      0.0   
1936754                 N         0           0.0          57.0     18.0   
1936755                 N         0           1.0           0.0     19.0   
1936756                 N         0           NaN           NaN      NaN   
1936757                 N         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
0                 NaN                NaN  
1                 NaN                NaN  
2                 NaN                NaN  
3                 0.0               32.0  
4                 NaN                NaN  
...               ...                ...  
1936753           0.0               22.0  
1936754           0.0                0.0  
1936755           0.0               79.0  
1936756           NaN                NaN  
1936757           NaN                NaN  

[1936758 rows x 30 columns]

# Brief description of the columns and their types and units

**Year** :
Year of the flight (stored as factor).

**Month** :
Month of the flight (stored as factor).

**DayOfMonth** :
Day of the Month (1 to 31, stored as integer).

**DayOfWeek** :
Day of the Week (stored as factor).

**DepTime** :
Actual Departure Time (stored as float, local time in hhmm)

**CRSDepTime** :
Computer Reservation System (scheduled) Departure Time (stored as float).

**ArrTime** :
Actual Arrival Time (stored as float, local time in hhmm)

**CRSArrTime** :
Computer Reservation System (scheduled) Arrival Time (stored as float).

**UniqueCarrier** :
Carrier ID (stored as factor).

**FlightNum** :
Flight Number (stored as factor).

**TailNum** :
Plane's Tail Number (stored as factor).

**ActualElapsedTime** :
Actual Elapsed Time of the flight, in minutes (stored as integer).

**CRSElapsedTime** :
Computer Reservation System (scheduled) Elapsed Time of the flight, in minutes (stored as integer).

**AirTime** :
Airborne Time for the flight, in minutes (stored as integer).

**ArrDelay** :
Arrival Delay, in minutes (stored as integer).

**DepDelay** :
Departure Delay, in minutes (stored as integer).

**Origin** :
Originating airport code (stored as factor).

**Dest** :
Destination airport code (stored as factor).

**Distance** :
Flight Distance between airports (stored as integer, distance in miles = 1,60934 km)

**TaxiIn** :
Taxi time from wheels down to arrival at the gate, in minutes (stored as integer).

**TaxiOut** :
Taxi time from departure from the gate to wheels up, in minutes (stored as integer).

**Cancelled** :
Cancellation status (stored as logical, 1 = canceled)

**CancellationCode** :
Cancellation code, if applicable (stored as factor) A = Carrier, B = Weather, C = National Air System, D = Security

**Diverted** :
Diversion status (stored as logical, 1 = diverted)

**CarrierDelay** :
Delay, in minutes, attributable to the carrier (stored integer).

**WeatherDelay** :
Delay, in minutes, attributable to weather factors (stored as integer).

**NASDelay** :
Delay, in minutes, attributable to the National Aviation System (stored as integer).

**SecurityDelay** :
Delay, in minutes, attributable to security factors (stored as integer).

**LateAircraftDelay** :
Delay, in minutes, attributable to late-arriving aircraft (stored as integer).

CSVs don't have indexes like our DataFrames, so all we need to do is just designate the index_col when reading


In [4]:
# Reading data from CSVs
DelayedFlights_df = pd.read_csv('DelayedFlights.csv', index_col=0)
# let's look again at the first 3 rows of the DataFrame
DelayedFlights_df[:3]

C:\Users\anaconda3\envs\myenv\lib\site-packages\numpy\lib\arraysetops.py:583: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)


Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  2008      1           3          4   2003.0        1955   2211.0   
1  2008      1           3          4    754.0         735   1002.0   
2  2008      1           3          4    628.0         620    804.0   

   CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0        2225            WN        335  ...    4.0      8.0          0   
1        1000            WN       3231  ...    5.0     10.0          0   
2         750            WN        448  ...    3.0     17.0          0   

   CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0                 N         0           NaN          NaN      NaN   
1                 N         0           NaN          NaN      NaN   
2                 N         0           NaN          NaN      NaN   

   SecurityDelay  LateAircraftDelay  
0            NaN                NaN  
1            NaN                NaN  
2            NaN                NaN  

[3 rows x 29 columns]

Here we are setting the index to be column zero. The vast majority of CSVs do not have an index column, so we usually won't have to worry about this step.

However, if we see that the file is broken, read_csv has a lot of options that will allow us to fix it. For example, if :

* Change the column separator to a ;
* Set the encoding to '_latin1_' (the default is '_utf8_')
* Parse the dates in the 'Date' column
* Tell it that our dates have the date first instead of the month first
* Set the index to be the 'Date' column

<!--- -->
[//]: # 

>``DelayedFlights_df = pd.read_csv('DelayedFlights.csv', sep=';', encoding='latin1', parse_dates=['Date'],dayfirst=True, index_col='Date')``


# Visualization of the data

The first thing we are going to do to visualize the new dataset (pandas dataframe) DelayedFlights_df is to print a few rows to have them as a visual reference. For this we are going to use the head method that returns the first 5 rows of the DataFrame:

In [5]:
# Preview the first 5 lines of the loaded data 
DelayedFlights_df.head()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  2008      1           3          4   2003.0        1955   2211.0   
1  2008      1           3          4    754.0         735   1002.0   
2  2008      1           3          4    628.0         620    804.0   
4  2008      1           3          4   1829.0        1755   1959.0   
5  2008      1           3          4   1940.0        1915   2121.0   

   CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0        2225            WN        335  ...    4.0      8.0          0   
1        1000            WN       3231  ...    5.0     10.0          0   
2         750            WN        448  ...    3.0     17.0          0   
4        1925            WN       3920  ...    3.0     10.0          0   
5        2110            WN        378  ...    4.0     10.0          0   

   CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0                 N         0           NaN          NaN      NaN   
1                 N         0           NaN          NaN      NaN   
2                 N         0           NaN          NaN      NaN   
4                 N         0           2.0          0.0      0.0   
5                 N         0           NaN          NaN      NaN   

   SecurityDelay  LateAircraftDelay  
0            NaN                NaN  
1            NaN                NaN  
2            NaN                NaN  
4            0.0               32.0  
5            NaN                NaN  

[5 rows x 29 columns]

By default the first five rows of the DataFrame are shown, but we could also pass a number: DelayedFlights_df.head(10) and the first ten rows would be shown.

In the same way to see the last five rows we use the .tail() method:

In [6]:
# Preview the last 5 lines of the loaded data 
DelayedFlights_df.tail()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
7009710  2008     12          13          6   1250.0        1220   1617.0   
7009717  2008     12          13          6    657.0         600    904.0   
7009718  2008     12          13          6   1007.0         847   1149.0   
7009726  2008     12          13          6   1251.0        1240   1446.0   
7009727  2008     12          13          6   1110.0        1103   1413.0   

         CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
7009710        1552            DL       1621  ...    9.0     18.0          0   
7009717         749            DL       1631  ...   15.0     34.0          0   
7009718        1010            DL       1631  ...    8.0     32.0          0   
7009726        1437            DL       1639  ...   13.0     13.0          0   
7009727        1418            DL       1641  ...    8.0     11.0          0   

         CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
7009710                 N         0           3.0          0.0      0.0   
7009717                 N         0           0.0         57.0     18.0   
7009718                 N         0           1.0          0.0     19.0   
7009726                 N         0           NaN          NaN      NaN   
7009727                 N         0           NaN          NaN      NaN   

         SecurityDelay  LateAircraftDelay  
7009710            0.0               22.0  
7009717            0.0                0.0  
7009718            0.0               79.0  
7009726            NaN                NaN  
7009727            NaN                NaN  

[5 rows x 29 columns]

This method also accepts a number "n" to print the last "n" rows.

Typically, after loading a dataset, we like to look at the first few rows to see the names of each column, the index and some example values for each of the rows.

# Get information about the data

One of the first commands you should run after loading data is .info(). This method gives us some essential details about the dataset, such as the number of rows and columns, the number of non-null values, what type of data is in each column, and how much memory the DataFrame is using.




In [7]:
# This method prints a summary of a DataFrame and returns None.
# Parameters: verbose=None, buf=None, max_cols=None, memory_usage=None, null_counts=None
DelayedFlights_df.info(show_counts=True) 

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1936758 entries, 0 to 7009727
Data columns (total 29 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   Year               1936758 non-null  int64  
 1   Month              1936758 non-null  int64  
 2   DayofMonth         1936758 non-null  int64  
 3   DayOfWeek          1936758 non-null  int64  
 4   DepTime            1936758 non-null  float64
 5   CRSDepTime         1936758 non-null  int64  
 6   ArrTime            1929648 non-null  float64
 7   CRSArrTime         1936758 non-null  int64  
 8   UniqueCarrier      1936758 non-null  object 
 9   FlightNum          1936758 non-null  int64  
 10  TailNum            1936753 non-null  object 
 11  ActualElapsedTime  1928371 non-null  float64
 12  CRSElapsedTime     1936560 non-null  float64
 13  AirTime            1928371 non-null  float64
 14  ArrDelay           1928371 non-null  float64
 15  DepDelay           1936758 non-n

Another very fast and useful attribute is .shape, which produces only a tuple of (rows, columns). We are going to use .shape as many times as we perform cleanup and transformation operations on the data. It is a quick and efficient way to know how many rows have been removed during the cleaning process.

In [8]:
# .shape return a tuple representing the dimensionality of the DataFrame.
DelayedFlights_df.shape

(1936758, 29)

So we have 1936758 rows and 30 columns in our DataFrame. Note that in our Delayed Flights data set we have some obvious missing values in the next columns:

* ArrTime
* TailNum
* ActualElapsedTime
* CRSElapsedTime
* AirTime
* ArrDelay
* TaxiIn
* TaxiOut
* CarrierDelay
* WeatherDelay
* NASDelay
* SecurityDelay
* LateAircraftDelay

where the number of non-zero values is less than the total number of rows 1936758. Next we will see how we treat them.

# Duplicate handling

This dataset has no duplicate rows, but it is always important to verify that duplicate rows are not being added.

To demonstrate this, let's simply duplicate our Delayed Flights DataFrame by adding it to itself:

In [9]:
# We are duplicating the data set
temp_df = DelayedFlights_df.append(DelayedFlights_df)
# Let's use .shape once again
temp_df.shape

(3873516, 29)

We use append() to get a copy of the DataFrame, this way we avoid affecting the original DataFrame and work on the copy stored in temp and not on the real data.

If we look at the output of .shape we can quickly see that the rows of our DataFrame have been duplicated.

Now let's try to remove the duplicates:

In [10]:
# Duplicate elimination
temp_df = temp_df.drop_duplicates()
# Let's use .shape once again
temp_df.shape

(1936756, 29)

As with append(), the drop_duplicates() method returns a copy of the DataFrame, but this time without duplicates. Calling .shape confirms that we are back to the original dataset with 1936756 rows.

Using the pandas argument with the keyword inplace=True we modify the DataFrame object directly avoiding having to keep assigning DataFrames to the same variable.

In [11]:
temp_df.drop_duplicates(inplace=True)
# display
temp_df.shape

(1936756, 29)

Now temp_df will already have the data transformed directly.

Another interesting argument for drop_duplicates() is keep, it has three options 

first: (default) Removes duplicates except the first occurrence.
last: Removes duplicates except the last occurrence.
false: Removes all duplicates.
In our case the default argument of keep is first. This means that if two rows are the same, pandas will remove the second row and keep the first. Using last has the opposite effect: the first row is removed.

keep, on the other hand, will remove all duplicates. If two rows are the same, both will be removed. See what happens with temp_df:

In [12]:
temp_df = DelayedFlights_df.append(DelayedFlights_df)  # make a new copy

temp_df.drop_duplicates(inplace=True, keep=False)
# display
temp_df.shape

(0, 29)

Since all rows were duplicates, keep=False deleted all rows, leaving the rows at zero. If one wonders why we would want to do that, one reason is that this option allows us to locate all duplicates in the data set. The following explains how to do this by means of conditional selections.

# Column cleaning

To facilitate the selection of data by column name we should spend a little time to simplify or rename names that are too convoluted or have symbols, words in upper and lower case, spaces or typos.

Next we are going to print the column names of the data set:

In [13]:
DelayedFlights_df.columns

Index(['Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime', 'CRSDepTime',
       'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum', 'TailNum',
       'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut',
       'Cancelled', 'CancellationCode', 'Diverted', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')

It is also very useful when trying to select data by column and we receive a Key Error, this .columns method allows us to simply copy and paste.

The .rename() method is used to rename certain or all columns using a dict.

In [14]:
DelayedFlights_df.rename(columns={
    'DayofMonth':'Day Of Month',
    'DayOfWeek':'Day Of Week',
    'DepTime':'Departure Time',
    'CRSDepTime':'CRS Departure Time',
    'ArrTime':'Arrival Time',
    'CRSArrTime':'CRS Arrival Time',
    'UniqueCarrier':'Unique Carrier',
    'FlightNum':'Flight Number',
    'TailNum':'Tail Number',
    'ActualElapsedTime':'Actual Elapsed Time',
    'CRSElapsedTime':'CRS Elapsed Time',
    'AirTime':'Air Time',
    'ArrDelay':'Arrival Delay',
    'DepDelay':'Departure Delay',
    'TaxiIn':'Taxi In',
    'TaxiOut':'Taxi Out',
    'CancellationCode':'Cancellation Code',
    'CarrierDelay':'Carrier Delay',
    'WeatherDelay':'Weather Delay',
    'NASDelay':'NAS Delay',
    'SecurityDelay':'Security Delay',
    'LateAircraftDelay':'Late Aircraft Delay'
    }, inplace = True)

DelayedFlights_df.columns

Index(['Year', 'Month', 'Day Of Month', 'Day Of Week', 'Departure Time',
       'CRS Departure Time', 'Arrival Time', 'CRS Arrival Time',
       'Unique Carrier', 'Flight Number', 'Tail Number', 'Actual Elapsed Time',
       'CRS Elapsed Time', 'Air Time', 'Arrival Delay', 'Departure Delay',
       'Origin', 'Dest', 'Distance', 'Taxi In', 'Taxi Out', 'Cancelled',
       'Cancellation Code', 'Diverted', 'Carrier Delay', 'Weather Delay',
       'NAS Delay', 'Security Delay', 'Late Aircraft Delay'],
      dtype='object')

To lowercase all names we will use a list of column names:

In [15]:
DelayedFlights_df.columns = [
'year',
'month',
'day_of_month',
'day_of_week',
'departure_time',
'crs_departure_time',
'arrival_time',
'crs_arrival_time',
'unique_carrier',
'flight_number',
'tail_number',
'actual_elapsed_time',
'crs_elapsed_time',
'air_time',
'arrival_delay',
'departure_delay',
'origin',
'dest',
'distance',
'taxi_in',
'taxi_out',
'cancelled',
'cancellation_code',
'diverted',
'carrier_delay',
'weather_delay',
'nas_delay',
'security_delay',
'late_aircraft_delay']

DelayedFlights_df.columns

Index(['year', 'month', 'day_of_month', 'day_of_week', 'departure_time',
       'crs_departure_time', 'arrival_time', 'crs_arrival_time',
       'unique_carrier', 'flight_number', 'tail_number', 'actual_elapsed_time',
       'crs_elapsed_time', 'air_time', 'arrival_delay', 'departure_delay',
       'origin', 'dest', 'distance', 'taxi_in', 'taxi_out', 'cancelled',
       'cancellation_code', 'diverted', 'carrier_delay', 'weather_delay',
       'nas_delay', 'security_delay', 'late_aircraft_delay'],
      dtype='object')

Another option but with less work and less prone to error is to make a list comprehension:

In [16]:
# change from uppercase to lowercase
DelayedFlights_df.columns = [col.lower() for col in DelayedFlights_df]
# replacing blank spaces with '_' 
DelayedFlights_df.columns =[column.replace(" ", "_") for column in DelayedFlights_df.columns]
# display
DelayedFlights_df.columns

Index(['year', 'month', 'day_of_month', 'day_of_week', 'departure_time',
       'crs_departure_time', 'arrival_time', 'crs_arrival_time',
       'unique_carrier', 'flight_number', 'tail_number', 'actual_elapsed_time',
       'crs_elapsed_time', 'air_time', 'arrival_delay', 'departure_delay',
       'origin', 'dest', 'distance', 'taxi_in', 'taxi_out', 'cancelled',
       'cancellation_code', 'diverted', 'carrier_delay', 'weather_delay',
       'nas_delay', 'security_delay', 'late_aircraft_delay'],
      dtype='object')

The use of list and dict are very useful when working with pandas and data in general.

It is a very good idea to lowercase, remove special characters and replace spaces with underscores if you are going to work with a data set for some time.

# How to work with missing values

When exploring data, we are likely to encounter missing or null values, which are essentially placeholders for non-existent values. Most commonly, we will encounter Python's None or NumPy's np.nan, each of which will be handled differently depending on the case.

There are two ways to deal with nulls:

1- Get rid of rows or columns with nulls.

2- Replace nulls with non-null values using a technique known as imputation.

First we must obtain the total number of null values for each column of our data set:

In [17]:
DelayedFlights_df.isnull()

year  month  day_of_month  day_of_week  departure_time  \
0        False  False         False        False           False   
1        False  False         False        False           False   
2        False  False         False        False           False   
4        False  False         False        False           False   
5        False  False         False        False           False   
...        ...    ...           ...          ...             ...   
7009710  False  False         False        False           False   
7009717  False  False         False        False           False   
7009718  False  False         False        False           False   
7009726  False  False         False        False           False   
7009727  False  False         False        False           False   

         crs_departure_time  arrival_time  crs_arrival_time  unique_carrier  \
0                     False         False             False           False   
1                     False         False             False           False   
2                     False         False             False           False   
4                     False         False             False           False   
5                     False         False             False           False   
...                     ...           ...               ...             ...   
7009710               False         False             False           False   
7009717               False         False             False           False   
7009718               False         False             False           False   
7009726               False         False             False           False   
7009727               False         False             False           False   

         flight_number  ...  taxi_in  taxi_out  cancelled  cancellation_code  \
0                False  ...    False     False      False              False   
1                False  ...    False     False      False              False   
2                False  ...    False     False      False              False   
4                False  ...    False     False      False              False   
5                False  ...    False     False      False              False   
...                ...  ...      ...       ...        ...                ...   
7009710          False  ...    False     False      False              False   
7009717          False  ...    False     False      False              False   
7009718          False  ...    False     False      False              False   
7009726          False  ...    False     False      False              False   
7009727          False  ...    False     False      False              False   

         diverted  carrier_delay  weather_delay  nas_delay  security_delay  \
0           False           True           True       True            True   
1           False           True           True       True            True   
2           False           True           True       True            True   
4           False          False          False      False           False   
5           False           True           True       True            True   
...           ...            ...            ...        ...             ...   
7009710     False          False          False      False           False   
7009717     False          False          False      False           False   
7009718     False          False          False      False           False   
7009726     False           True           True       True            True   
7009727     False           True           True       True            True   

         late_aircraft_delay  
0                       True  
1                       True  
2                       True  
4                      False  
5                       True  
...                      ...  
7009710                False  
7009717                False  
7009718                False  
7009726                 True  
70097

Using isnull() we will get for each cell of the DataFrame whether it is True or False depending on whether the cell is null or not.

To count the number of nulls in each column we use a function to sum them:

In [18]:
DelayedFlights_df.isnull().sum()

year                        0
month                       0
day_of_month                0
day_of_week                 0
departure_time              0
crs_departure_time          0
arrival_time             7110
crs_arrival_time            0
unique_carrier              0
flight_number               0
tail_number                 5
actual_elapsed_time      8387
crs_elapsed_time          198
air_time                 8387
arrival_delay            8387
departure_delay             0
origin                      0
dest                        0
distance                    0
taxi_in                  7110
taxi_out                  455
cancelled                   0
cancellation_code           0
diverted                    0
carrier_delay          689270
weather_delay          689270
nas_delay              689270
security_delay         689270
late_aircraft_delay    689270
dtype: int64

.isnull() is not very useful in isolation but it is useful when used as in this case with another method such as sum().

We see that the data has null values for:

| Column Name :       | Nulls :            |
|:--------------------|:-------------------|
|arrival_time:        |    7110            |
|tail_num:            |       5            |
|actual_elapsed_time: |    8387            |
|crs_elapsed_time:    |     198            |
|air_time:            |    8387            |
|arrival_delay:       |    8387            |
|taxi_in:             |    7110            |
|taxi_out:            |     455            |
|carrier_delay:       |  689270            |
|weather_delay:       |  689270            |
|nas_delay:           |  689270            | 
|security_delay:      |  689270            | 
|late_aircraft_delay: |  689270            |

# Null value removal

This decision requires a thorough understanding of the data and its context. Usually the removal of null data is performed only when a small amount of missing data is involved.

We will remove null data as follows:

In [19]:
# Removing null values
# This operation will delete any row with at least a single null value, but it will return a new DataFrame without altering the original one. 
DelayedFlights_df.dropna(inplace = True)

By this operation any row with at least one null value has been removed. Using the inplace=True method we get a new DataFrame that includes these changes. 

In this way we have removed a large number of rows from the dataset that contained null data. This operation produces a large amount of wasted data because there is perfectly valid data in the other columns of these same deleted rows. For this reason, we will now use another alternative called imputation.

This alternative consists of eliminating the columns with null values (by setting axis = 1) instead of eliminating the rows:

In [20]:
# Other than just dropping rows, you can also drop columns with null values by setting axis=1:
DelayedFlights_df.dropna(axis = 1, inplace = True)

In the data set, this operation would remove all columns with null values.

To verify the results of the operation we use the .shape method:

In [21]:
DelayedFlights_df.shape

(1247486, 29)

# Imputation

Imputation is a very useful technique for retaining valuable data that have null values. As it happens in this case the elimination of rows with null values means the elimination of too large a part of the data set. To avoid this we will impute each null value by another value such as the mean or median of that column because it is more convenient.

We will analyze the imputation of the null values in the arrival_delay column. To do this we will extract that column in its own variable:

In [22]:
arrival_delay_col = DelayedFlights_df['arrival_delay']

arrival_delay_col

4          34.0
6          57.0
11         80.0
16         15.0
18         16.0
           ... 
7009705    64.0
7009709    17.0
7009710    25.0
7009717    75.0
7009718    99.0
Name: arrival_delay, Length: 1247486, dtype: float64

The way to select columns in a DataFrame is done using square brackets. The dict keys in a DataFrame are the column names, so when selecting columns in a DataFrame, we use square brackets like when accessing a Python dictionary.

So now arrival_delay_col contains the Series:

In [23]:
arrival_delay_col.head()

4     34.0
6     57.0
11    80.0
16    15.0
18    16.0
Name: arrival_delay, dtype: float64

The format of the Series is similar to that of a DataFrame.

Let's replace the missing values in arrival_delay with the mean of the Series. To do this, we will first calculate the mean value:


In [24]:
arrival_delay_mean = arrival_delay_col.mean()

arrival_delay_mean

63.291035731062316

Now let's fill the nulls of the series with the mean using fillna():

In [25]:
arrival_delay_col.fillna(arrival_delay_mean, inplace = True)

All nulls in the arrivaldelay series have been replaced by the column mean. Using inplace = True we have directly updated the original DataFrame DelayedFlights_df:

In [26]:
DelayedFlights_df.isnull().sum()

year                   0
month                  0
day_of_month           0
day_of_week            0
departure_time         0
crs_departure_time     0
arrival_time           0
crs_arrival_time       0
unique_carrier         0
flight_number          0
tail_number            0
actual_elapsed_time    0
crs_elapsed_time       0
air_time               0
arrival_delay          0
departure_delay        0
origin                 0
dest                   0
distance               0
taxi_in                0
taxi_out               0
cancelled              0
cancellation_code      0
diverted               0
carrier_delay          0
weather_delay          0
nas_delay              0
security_delay         0
late_aircraft_delay    0
dtype: int64

Replacing an entire column with the same value is easiest, but we could also try to make replacements at a higher granularity according to some classification if that is the case. For example, we could find the mean arrival delay for each of the delay reasons that occur (carrier_delay, weather_delay, nas_delay, security_delay, late_aircraft_delay) and replace the nulls for each reason with the mean for that reason.

Let us now look at more ways to examine and understand the data set.

# Understanding the variables

We will use .describe() in the DataFrame to get a summary of the distribution of the continuous variables, for each of the columns we will see a few statistics.

In [27]:
#
pd.options.display.max_columns = 10
# removing null values to avoid errors 
DelayedFlights_df.dropna(inplace = True)

# percentile list
perc =[.20, .40, .60, .80]
  
# list of dtypes to include
include =['object', 'float', 'int']
  
# calling describe method
desc = DelayedFlights_df.describe(percentiles = perc, include = include)
#desc = DelayedFlights_df.describe()

# set option to format the output
pd.set_option('display.float_format', lambda x: '%.5f' % x)
  
# display
desc

departure_time  arrival_time unique_carrier tail_number  \
count    1247486.00000 1247486.00000        1247486     1247486   
unique             NaN           NaN             20        5349   
top                NaN           NaN             WN      N17175   
freq               NaN           NaN         203559         686   
mean        1558.83193    1616.74890            NaN         NaN   
std          454.33003     583.94163            NaN         NaN   
min            1.00000       1.00000            NaN         NaN   
20%         1139.00000    1223.00000            NaN         NaN   
40%         1455.00000    1609.00000            NaN         NaN   
50%         1618.00000    1737.00000            NaN         NaN   
60%         1736.00000    1857.00000            NaN         NaN   
80%         2000.00000    2122.00000            NaN         NaN   
max         2400.00000    2400.00000            NaN         NaN   

        actual_elapsed_time  ...  carrier_delay  weather_delay     nas_delay  \
count         1247486.00000  ...  1247486.00000  1247486.00000 1247486.00000   
unique                  NaN  ...            NaN            NaN           NaN   
top                     NaN  ...            NaN            NaN           NaN   
freq                    NaN  ...            NaN            NaN           NaN   
mean              135.37789  ...       19.17943        3.70336      15.02162   
std                72.29636  ...       43.54624       21.49153      33.83308   
min                14.00000  ...        0.00000        0.00000       0.00000   
20%                76.00000  ...        0.00000        0.00000       0.00000   
40%               103.00000  ...        0.00000        0.00000       0.00000   
50%               118.00000  ...        2.00000        0.00000       2.00000   
60%               136.00000  ...        9.00000        0.00000       6.00000   
80%               182.00000  ...       27.00000        0.00000      20.00000   
max              1114.00000  ...     2436.00000     1352.00000    1357.00000   

        security_delay late_aircraft_delay  
count    1247486.00000       1247486.00000  
unique             NaN                 NaN  
top                NaN                 NaN  
freq               NaN                 NaN  
mean           0.09014            25.29649  
std            2.02272            42.05489  
min            0.00000             0.00000  
20%            0.00000             0.00000  
40%            0.00000             0.00000  
50%            0.00000             8.00000  
60%            0.00000            17.00000  
80%            0.00000            42.00000  
max          392.00000          1316.00000  

[13 rows x 19 columns]

Note that it only gives statistics for the numerical columns. It is also very useful to know when the numbers are continuous to determine what type of chart can be used to represent the data visually.

Let's review what each of these statistics means:

+ **Count**: This is the number of rows that have a value. In our case, each flight has a value for each of the columns, so the value is 1247486 (the total number of flights).<br/>
+ **Mean**: Recall that the mean is the standard average.<br/>
+ **Std**: This is short for standard deviation. This is a measure of how dispersed the data is.<br/>
+ **Min**: The smallest value.<br/>
+ **25%**: The 25th percentile.<br/>
+ **50%**: The 50th percentile, also known as the median.<br/>
+ **75%**: The 75th percentile.<br/>
+ **Max**: The largest value.

We can also use .describe() on a categorical variable to get the row count, unique category count, top category, and top category frequency:

In [28]:
DelayedFlights_df['unique_carrier'].describe()

count     1247486
unique         20
top            WN
freq       203559
Name: unique_carrier, dtype: object

This tells us that the unique_carrier column has 20 unique values, the highest value being WN, which appears 203559 times (frequency).

We can obtain the frequency of all the values in a column as follows: 

In [29]:
#DelayedFlights_df['uniquecarrier'].value_counts().head(10)
#DelayedFlights_df['flightnum'].value_counts().head(10)
DelayedFlights_df['tail_number'].value_counts().head(10)


N17175    686
N77278    679
N87353    678
N37208    665
N37342    660
N27172    647
N17156    636
N75995    635
N651ML    633
N27173    631
Name: tail_number, dtype: int64

# Relationships between continuous variables

We will perform the correlation between continuous variables as follows:

In [30]:
DelayedFlights_df.corr()

year    month  day_of_month  day_of_week  departure_time  \
year                  NaN      NaN           NaN          NaN             NaN   
month                 NaN  1.00000       0.06703      0.00929        -0.00876   
day_of_month          NaN  0.06703       1.00000      0.02595         0.00340   
day_of_week           NaN  0.00929       0.02595      1.00000         0.02141   
departure_time        NaN -0.00876       0.00340      0.02141         1.00000   
crs_departure_time    NaN -0.01344       0.00359      0.02781         0.84026   
arrival_time          NaN -0.00196       0.00341      0.00727         0.37323   
crs_arrival_time      NaN -0.00182       0.00395      0.01727         0.71422   
flight_number         NaN  0.00255      -0.00490     -0.01001        -0.03174   
actual_elapsed_time   NaN  0.00398      -0.00001      0.00454        -0.05682   
crs_elapsed_time      NaN  0.00661       0.00056      0.00620        -0.04148   
air_time              NaN  0.00011       0.00036      0.00584        -0.05862   
arrival_delay         NaN  0.01234       0.00907      0.01190         0.09398   
departure_delay       NaN  0.01540       0.00987      0.01380         0.11581   
distance              NaN  0.00583      -0.00021      0.00899        -0.05488   
taxi_in               NaN  0.02390      -0.00136      0.01196        -0.02512   
taxi_out              NaN  0.00815      -0.00104     -0.00853         0.00242   
cancelled             NaN      NaN           NaN          NaN             NaN   
diverted              NaN      NaN           NaN          NaN             NaN   
carrier_delay         NaN  0.00042      -0.00095      0.01021        -0.05195   
weather_delay         NaN  0.00661       0.00090      0.00564         0.00530   
nas_delay             NaN  0.01144       0.00564     -0.00663         0.02253   
security_delay        NaN -0.00307      -0.00031      0.00467        -0.01527   
late_aircraft_delay   NaN  0.00495       0.00910      0.00884         0.16945   

                     ...  carrier_delay  weather_delay  nas_delay  \
year                 ...            NaN            NaN        NaN   
month                ...        0.00042        0.00661    0.01144   
day_of_month         ...       -0.00095        0.00090    0.00564   
day_of_week          ...        0.01021        0.00564   -0.00663   
departure_time       ...       -0.05195        0.00530    0.02253   
crs_departure_time   ...       -0.10734       -0.00933   -0.03270   
arrival_time         ...       -0.08398       -0.02987    0.03239   
crs_arrival_time     ...       -0.09616       -0.00503    0.00923   
flight_number        ...        0.05571        0.06748    0.01540   
actual_elapsed_time  ...        0.00996        0.00407    0.17714   
crs_elapsed_time     ...        0.02659       -0.01647    0.04696   
air_time             ...        0.01955       -0.01832    0.06279   
arrival_delay        ...        0.50178        0.26460    0.38761   
departure_delay      ...        0.52804        0.24447    0.23398   
distance             ...        0.02844       -0.02382    0.01811   
taxi_in              ...       -0.02198        0.02566    0.22983   
taxi_out             ...       -0.02876        0.08324    0.42799   
cancelled            ...            NaN            NaN        NaN   
diverted             ...            NaN            NaN        NaN   
carrier_delay        ...        1.00000       -0.06895   -0.11953   
weather_delay        ...       -0.06895        1.00000   -0.00255   
nas_delay            ...       -0.11953       -0.00255    1.00000   
security_delay       ...       -0.01854       -0.00751   -0.01172   
late_aircraft_delay  ...       -0.17829       -0.05497   -0.11890   

                     security_delay  late_aircraft_delay  
year                            NaN                  NaN  
month                      -0.00307              0.00495  
day_of_month               -0.00031              0.00910  
day_of_week                 0.004

Correlation tables are a numerical representation of the bivariate relationships in the data set.

Positive numbers indicate a positive correlation - one goes up and one goes up - and negative numbers represent a reverse correlation - one goes up and one goes down. 1.0 indicates a perfect correlation.

Examining bivariate relationships is useful when you have an outcome or dependent variable in mind and want to see the characteristics most correlated with the increase or decrease in the outcome.

In [31]:
# only view these two variables
DelayedFlights_df[['arrival_delay', 'departure_delay']].corr()

arrival_delay  departure_delay
arrival_delay          1.00000          0.95032
departure_delay        0.95032          1.00000

If we look at the first row and the first column of the correlation matrix table, we see that the rank has a perfect correlation with itself, which is evident. On the other hand, the correlation between arrival_delay and departure_delay is 0.95032, which indicates a strong positive correlation between arrival_delay and departure_delay. It is evident that there is a relationship between the two variables.

# DataFrame slicing, selecting, extracting

It is important to keep in mind that, although many methods are the same, DataFrames and Series have different attributes, so we must make sure we know which type we are working with or we will receive attribute errors.

Let's first look at working with columns.

## By column

Often we only want to extract some of the columns we have in our dataset. To extract a single column, we use square brackets and the column name.

In the following example, we will extract only the arrivaldelay column in this way:

In [32]:
arrivaldelay_col = DelayedFlights_df['arrival_delay']

type(arrivaldelay_col)

pandas.core.series.Series

This will return a Series. To extract a column as a DataFrame, we need to pass a list of column names. In this case it is a single column:

In [33]:
arrivaldelay_col = DelayedFlights_df[['arrival_delay']]

type(arrivaldelay_col)

pandas.core.frame.DataFrame

# Selecting Multiple Columns

We can also select multiple columns from our original DataFrame, creating a smaller DataFrame.
We're going to select just the unique_carrier, flight_number, departure_delay and arrivel_delay columns from our original DataFrame.
We put these values in a list as follows:

``['unique_carrier', 'flight_number', 'departure_delay', 'arrivel_delay']``

Now we use that list inside of the bracket notation DelayedFlights_df\[...\] When printing a large DataFrame that’s too big to display, you can use the head method to print just the first 5 rows.

Since it's just a list, adding another column name is easy:

In [34]:
subset = DelayedFlights_df[['unique_carrier', 'flight_number', 'arrival_delay', 'departure_delay']]

subset.head()

unique_carrier  flight_number  arrival_delay  departure_delay
4              WN           3920       34.00000         34.00000
6              WN            509       57.00000         67.00000
11             WN           1333       80.00000         94.00000
16             WN            675       15.00000         27.00000
18             WN              4       16.00000         28.00000

In [35]:
DelayedFlights_df.rename(columns={'day_of_month' : 'day'}, inplace=True)

data_set_1 = DelayedFlights_df[['year', 'month', 'day']]
data_set_2 = DelayedFlights_df['departure_time']

date_time_df = pd.DataFrame()

date_time_df['date'] = pd.to_datetime(data_set_1, dayfirst=False, errors='coerce')

date_time_df['time'] = pd.to_timedelta(data_set_2, unit='m')

date_time_df['date/time'] = pd.to_datetime(date_time_df.pop('date')) + pd.to_timedelta(date_time_df.pop('time'))

#date_time_df.rename(columns={'flight_date_time':'DATE/TIME'})

date_time_df.head()


date/time
4  2008-01-04 06:29:00
6  2008-01-04 08:17:00
11 2008-01-04 03:24:00
16 2008-01-04 00:12:00
18 2008-01-03 22:03:00

In [36]:
dw_mapping = {
    0: 'Monday', 
    1: 'Tuesday', 
    2: 'Wednesday', 
    3: 'Thursday', 
    4: 'Friday',
    5: 'Saturday', 
    6: 'Sunday'
}

day_of_week_name_df = pd.DataFrame()

day_of_week_name_df['week_day_name'] = date_time_df['date/time'].dt.weekday.map(dw_mapping)

day_of_week_name_df.head()

week_day_name
4         Friday
6         Friday
11        Friday
16        Friday
18      Thursday

In [37]:
day_of_week_name_df.value_counts().head(10)

week_day_name
Friday           200001
Monday           199310
Saturday         196065
Tuesday          177062
Wednesday        169748
Thursday         168916
Sunday           136384
dtype: int64

Now we will see how to get the data by rows.

# By rows

For rows, we have two options:

* .loc - locate by name

* .iloc- locate by numerical index

Recall that we are still indexing by the date and time of the flight, so to use .loc we give it the date of the flight:

In [38]:
flight = date_time_df.loc[:516,['date/time']]
# display
flight.head()

date/time
4  2008-01-04 06:29:00
6  2008-01-04 08:17:00
11 2008-01-04 03:24:00
16 2008-01-04 00:12:00
18 2008-01-03 22:03:00

Columns in a Pandas DataFrame can be accessed by their index in the DataFrame rather than by their column name. This can be useful if the columns in a DataFrame are unnamed or if there is more than one column with the same name.

In [39]:
first_10_columns = DelayedFlights_df.iloc[:3,:10]
first_10_columns

year  month  day  day_of_week  departure_time  crs_departure_time  \
4   2008      1    3            4      1829.00000                1755   
6   2008      1    3            4      1937.00000                1830   
11  2008      1    3            4      1644.00000                1510   

    arrival_time  crs_arrival_time unique_carrier  flight_number  
4     1959.00000              1925             WN           3920  
6     2037.00000              1940             WN            509  
11    1845.00000              1725             WN           1333

In [40]:
first_10_columns = DelayedFlights_df.iloc[:3,10:20]
first_10_columns

tail_number  actual_elapsed_time  crs_elapsed_time  air_time  \
4       N464WN             90.00000          90.00000  77.00000   
6       N763SW            240.00000         250.00000 230.00000   
11      N334SW            121.00000         135.00000 107.00000   

    arrival_delay  departure_delay origin dest  distance  taxi_in  
4        34.00000         34.00000    IND  BWI       515  3.00000  
6        57.00000         67.00000    IND  LAS      1591  3.00000  
11       80.00000         94.00000    IND  MCO       828  6.00000

In [41]:
first_10_columns = DelayedFlights_df.iloc[:3,20:30]
first_10_columns

taxi_out  cancelled cancellation_code  diverted  carrier_delay  \
4   10.00000          0                 N         0        2.00000   
6    7.00000          0                 N         0       10.00000   
11   8.00000          0                 N         0        8.00000   

    weather_delay  nas_delay  security_delay  late_aircraft_delay  
4         0.00000    0.00000         0.00000             32.00000  
6         0.00000    0.00000         0.00000             47.00000  
11        0.00000    0.00000         0.00000             72.00000

In [ ]:
DelayedFlights_df['arrival_delay'].plot()

In [ ]:
#per1 = pd.date_range(start ='1-1-2008', end ='30-12-2008', freq = 'D')
#per1 = pd.date_range(start ='1-1-2008', end ='30-12-2008', periods = 10, tz ='Asia/Tokyo')
per1 = pd.period_range(start ='2008-01-01', end ='2008-12-30', freq ='M')

for val in per1:
    print(val)

In [ ]:
#departures_df.plot()
plt.scatter(DelayedFlights_df['departure_time'], DelayedFlights_df['departure_delay'])

We can also plot all the columns just as easily. We'll make it a little bigger, too. You can see that it's more squished together, but all the bike paths behave basically the same – if it's a bad day for cyclists, it's a bad day everywhere.

## Create a Histogram of flight distance

In [ ]:
plt.hist(x=sample_df['distance'], bins = 'auto')
plt.xlabel('distance')
plt.ylabel('frequency')

## Plot elapsed time vs distance

In [ ]:
# Generate scatterplot of flight distance vs time
plt.scatter('distance', 'actual_elapsed_time', data = sample_df)
plt.xlabel('distance')
plt.ylabel('time')

## Plot arrival delay over time

In [ ]:
plt.scatter('month', 'arrival_delay', data = sample_df)
plt.xlabel('Month')
plt.ylabel('Arrival Delay')

## Create a histogram of arrival delay

In [ ]:
# Set missing values to 0
arvl_delay = sample_df['arrival_delay'].fillna(0)

plt.hist(x=arvl_delay, bins='auto')
plt.xlabel('Distance')
plt.ylabel('Frequency')

## List Top 5 airlines which occur delays most in 2008.

In [ ]:
Top5delay_df = DelayedFlights_df[['unique_carrier','departure_delay']].copy()
sorted_df = Top5delay_df.sort_values(by=['departure_delay'], ascending = False, inplace = False).groupby('unique_carrier').head(1)
sorted_df.head()

In [ ]:
sorted_df.plot(x = 'unique_carrier', y = 'departure_delay', kind="bar", figsize=(15,7))

In [ ]:
# Sort pandas dataframe by a specific column's values
sorted_df = DelayedFlights_df.sort_values(by=['departure_delay'],
                                                        axis=0,
                                                        ascending=False,
                                                        inplace=False,
                                                        kind='quicksort',
                                                        na_position='last',
                                                        ignore_index=False,
                                                        key=None)
sorted_df.head()

In [ ]:
# Take a smaller sample of the dataset for easier plotting
sample_df = sorted_df.sample(n=5)

ax = sample_df.plot(x='UniqueCarrier', y='DepDelay', kind="bar", figsize=(15,7))
ax.set_title("The Top 5 Airlines that Occur Departure Delays Most Frequently in 2008")
ax.set_ylabel('Count')


# Exercici 2 :

Fes un informe complet del data set.

   * Resumeix estadísticament les columnes d’interès
   * Troba quantes dades faltants hi ha per columna
   * Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...)
   * Taula de les aerolínies amb més endarreriments acumulats
   * Quins són els vols més llargs? I els més endarrerits?
   * Etc.

In [42]:
# Create a new subset of the DataFrame
data_set_df = DelayedFlights_df[['dest','flight_number','distance','air_time','unique_carrier','departure_time','departure_delay','arrival_time','arrival_delay','carrier_delay']]
# Create a DataFrame
new_data_set_df = pd.DataFrame(data_set_df)

In [43]:
# set option to format the output
pd.set_option('display.float_format', lambda x: '%.1f' % x)
# display a summary of the statistics 
new_data_set_df.describe()

flight_number  distance  air_time  departure_time  departure_delay  \
count      1247486.0 1247486.0 1247486.0       1247486.0        1247486.0   
mean          2276.5     741.6     107.4          1558.8             59.7   
std           1997.5     559.4      68.2           454.3             59.9   
min              1.0      11.0       0.0             1.0              6.0   
25%            637.0     334.0      58.0          1232.0             24.0   
50%           1589.0     595.0      90.0          1618.0             41.0   
75%           3676.0     972.0     136.0          1924.0             75.0   
max           9741.0    4962.0    1091.0          2400.0           2467.0   

       arrival_time  arrival_delay  carrier_delay  
count     1247486.0      1247486.0      1247486.0  
mean         1616.7           63.3           19.2  
std           583.9           60.8           43.5  
min             1.0           15.0            0.0  
25%          1326.0           26.0            0.0  
50%          1737.0           43.0            2.0  
75%          2048.0           79.0           21.0  
max          2400.0         2461.0         2436.0

   * Troba quantes dades faltants hi ha per columna

In [44]:
# Count NaN in each column of a DataFrame
print("NaN in each columns" , new_data_set_df.isnull().sum(), sep = '\n')

NaN in each columns
dest               0
flight_number      0
distance           0
air_time           0
unique_carrier     0
departure_time     0
departure_delay    0
arrival_time       0
arrival_delay      0
carrier_delay      0
dtype: int64


In [45]:
import operator
import numpy as np
np.random.seed(0)
df = pd.DataFrame(np.random.rand(3) * 1000, columns=['minutes'])
fmt = operator.methodcaller('strftime', '%H:%M:%S')
df['HH:MM:SS'] = pd.to_datetime(df['minutes'], unit='m').map(fmt)
df

minutes  HH:MM:SS
0    548.8  09:08:48
1    715.2  11:55:11
2    602.8  10:02:45

In [46]:
data_set_HMS_df = pd.DataFrame()
fmt = operator.methodcaller('strftime', '%H:%M:%S')
data_set_HMS_df['air_time_HMS'] = pd.to_datetime(new_data_set_df['air_time'], unit='m').map(fmt)
data_set_HMS_df['departure_time_HMS'] = pd.to_datetime(new_data_set_df['departure_time'], unit='m').map(fmt)
data_set_HMS_df['departure_delay_HMS'] = pd.to_datetime(new_data_set_df['departure_delay'], unit='m').map(fmt)
data_set_HMS_df['arrival_time_HMS'] = pd.to_datetime(new_data_set_df['arrival_time'], unit='m').map(fmt)
data_set_HMS_df['arrival_delay_HMS'] = pd.to_datetime(new_data_set_df['arrival_delay'], unit='m').map(fmt)
data_set_HMS_df['carrier_delay_HMS'] = pd.to_datetime(new_data_set_df['carrier_delay'], unit='m').map(fmt)
data_set_fmt_df = pd.DataFrame(data_set_HMS_df)
data_set_fmt_df.head()

air_time_HMS departure_time_HMS departure_delay_HMS arrival_time_HMS  \
4      01:17:00           06:29:00            00:34:00         08:39:00   
6      03:50:00           08:17:00            01:07:00         09:57:00   
11     01:47:00           03:24:00            01:34:00         06:45:00   
16     03:33:00           00:12:00            00:27:00         03:20:00   
18     01:50:00           22:03:00            00:28:00         01:26:00   

   arrival_delay_HMS carrier_delay_HMS  
4           00:34:00          00:02:00  
6           00:57:00          00:10:00  
11          01:20:00          00:08:00  
16          00:15:00          00:03:00  
18          00:16:00          00:00:00

 * Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...)

In [47]:
mean_velocity_df = pd.DataFrame()
# Conversion factor from miles to km (1.60934) and minutes to hour (60)
mean_velocity_df['mean_velocity'] = (new_data_set_df['distance'] * 1.60934 * 60 / new_data_set_df['air_time'])
# Print the mean velocity in km/h
mean_velocity_df['mean_velocity']

4         645.8
6         667.9
11        747.2
16        675.0
18        735.6
           ... 
7009705   641.8
7009709   540.0
7009710   729.0
7009717   595.5
7009718   545.3
Name: mean_velocity, Length: 1247486, dtype: float64

In [48]:
pd.options.display.max_columns = 12
flight_data_df = new_data_set_df[['dest','flight_number','distance','unique_carrier']]
df_obj = pd.concat([date_time_df, day_of_week_name_df, flight_data_df, data_set_fmt_df,mean_velocity_df], axis=1)
df_obj.set_index('date/time', inplace = True)
df = df_obj.set_axis(['DAY','DESTINATION','FLIGHT NUMBER','DISTANCE','AIRLINE','FLIGHT TIME','DEPARTURE','DEP. DELAY','ARRIVAL','ARR. DELAY','AIRLINE DELAY','MEAN VELOCITY'], axis = 1)
df.head()

DAY DESTINATION  FLIGHT NUMBER  DISTANCE AIRLINE  \
date/time                                                                    
2008-01-04 06:29:00    Friday         BWI           3920       515      WN   
2008-01-04 08:17:00    Friday         LAS            509      1591      WN   
2008-01-04 03:24:00    Friday         MCO           1333       828      WN   
2008-01-04 00:12:00    Friday         PHX            675      1489      WN   
2008-01-03 22:03:00  Thursday         TPA              4       838      WN   

                    FLIGHT TIME DEPARTURE DEP. DELAY   ARRIVAL ARR. DELAY  \
date/time                                                                   
2008-01-04 06:29:00    01:17:00  06:29:00   00:34:00  08:39:00   00:34:00   
2008-01-04 08:17:00    03:50:00  08:17:00   01:07:00  09:57:00   00:57:00   
2008-01-04 03:24:00    01:47:00  03:24:00   01:34:00  06:45:00   01:20:00   
2008-01-04 00:12:00    03:33:00  00:12:00   00:27:00  03:20:00   00:15:00   
2008-01-03 22:03:00    01:50:00  22:03:00   00:28:00  01:26:00   00:16:00   

                    AIRLINE DELAY  MEAN VELOCITY  
date/time                                         
2008-01-04 06:29:00      00:02:00          645.8  
2008-01-04 08:17:00      00:10:00          667.9  
2008-01-04 03:24:00      00:08:00          747.2  
2008-01-04 00:12:00      00:03:00          675.0  
2008-01-03 22:03:00      00:00:00          735.6

In [49]:
delay_bool_df = new_data_set_df['departure_delay'] > 0
delay_df = new_data_set_df[delay_bool_df]
import operator
fmt = operator.methodcaller('strftime', '%H:%M:%S')
delay_df['DELAY'] = pd.to_datetime(new_data_set_df['departure_delay'], unit='m').map(fmt)
df = delay_df[['dest','unique_carrier','flight_number','DELAY']]
df = df.set_axis(['DESTINATION','AIRLINE','FLIGHT NUMBER','DELAY'], axis = 1)
df.set_index('DESTINATION', inplace=True)
df

AIRLINE  FLIGHT NUMBER     DELAY
DESTINATION                                 
BWI              WN           3920  00:34:00
LAS              WN            509  01:07:00
MCO              WN           1333  01:34:00
PHX              WN            675  00:27:00
TPA              WN              4  00:28:00
...             ...            ...       ...
PBI              DL           1616  00:51:00
ATL              DL           1620  00:32:00
ATL              DL           1621  00:30:00
ATL              DL           1631  00:57:00
IAH              DL           1631  01:20:00

[1247486 rows x 3 columns]

In [52]:
temp_df = pd.DataFrame()
temp_df['departure_HMS'] = pd.to_timedelta(new_data_set_df['departure_time'], unit='m')
temp_df['arrival_HMS']   = pd.to_timedelta(new_data_set_df['arrival_time'], unit='m')
temp_df['delay_HMS']     = temp_df['arrival_HMS'] - temp_df['departure_HMS']
temp_df.set_index('departure_HMS', inplace=True)
temp_df

arrival_HMS       delay_HMS
departure_HMS                                  
1 days 06:29:00 1 days 08:39:00 0 days 02:10:00
1 days 08:17:00 1 days 09:57:00 0 days 01:40:00
1 days 03:24:00 1 days 06:45:00 0 days 03:21:00
1 days 00:12:00 1 days 03:20:00 0 days 03:08:00
0 days 22:03:00 1 days 01:26:00 0 days 03:23:00
...                         ...             ...
0 days 15:21:00 0 days 18:32:00 0 days 03:11:00
1 days 01:52:00 1 days 04:55:00 0 days 03:03:00
0 days 20:50:00 1 days 02:57:00 0 days 06:07:00
0 days 10:57:00 0 days 15:04:00 0 days 04:07:00
0 days 16:47:00 0 days 19:09:00 0 days 02:22:00

[1247486 rows x 2 columns]

  * Taula de les aerolínies amb més endarreriments acumulats

In [50]:
# filtering with query method
# CarrierDelay : Delay, in minutes, attributable to the carrier (stored integer).
DelayedFlights_df.query('carrier_delay == True')
# display
DelayedFlights_df.head()

year  month  day  day_of_week  departure_time  crs_departure_time  ...  \
4   2008      1    3            4          1829.0                1755  ...   
6   2008      1    3            4          1937.0                1830  ...   
11  2008      1    3            4          1644.0                1510  ...   
16  2008      1    3            4          1452.0                1425  ...   
18  2008      1    3            4          1323.0                1255  ...   

    diverted  carrier_delay weather_delay  nas_delay security_delay  \
4          0            2.0           0.0        0.0            0.0   
6          0           10.0           0.0        0.0            0.0   
11         0            8.0           0.0        0.0            0.0   
16         0            3.0           0.0        0.0            0.0   
18         0            0.0           0.0        0.0            0.0   

    late_aircraft_delay  
4                  32.0  
6                  47.0  
11                 72.0  
16                 12.0  
18                 16.0  

[5 rows x 29 columns]

In [51]:
Top5CarrierDelay_df = DelayedFlights_df[['unique_carrier','carrier_delay']].copy()
sorted_df = Top5CarrierDelay_df.sort_values(by=['carrier_delay'], ascending = False, inplace = False).groupby('unique_carrier').head(1)
sorted_df = sorted_df.set_axis(['AIRLINE','DELAY'], axis = 1)
sorted_df.set_index('AIRLINE', inplace=True)
sorted_df.head()

DELAY
AIRLINE       
NW      2436.0
MQ      1707.0
AA      1312.0
FL      1175.0
9E      1127.0

* Quins són els vols més llargs? I els més endarrerits?

| Origin | Destination | Distance |
|:---:|:---:|:---:|
| EWR	| HNL | 4962 |
| HNL	| EWR | 4962 |
| ATL	| HNL | 4502 |
| ORD	| HNL | 4243 |
| KOA	| ORD | 4213 |

In [55]:
set_data_df = DelayedFlights_df[['origin','dest','distance']].copy()
#show_data_df = set_data_df.query("origin == dest")
set_data_df.sample(n = 10)
#DelayedFlights_df['origin']== DelayedFlights_df['dest']
#sorted_flights_df = set_data_df.sort_values(by=['distance'], ascending = True, inplace = False).groupby('distance').max()
#groupby_data_df = set_data_df.groupby('distance').max()
#groupby_data_df
#sorted_flights_df = groupby_data_df.sort_values(by=['distance'], ascending = False, inplace = False)
#sorted_flights_df.head(10)

origin dest  distance
4794521    JFK  AUS      1522
826445     ORD  BOS       867
3278081    CLT  PHX      1774
3470189    TYS  DTW       443
6566864    IAH  BNA       657
950010     ROC  ATL       749
4162428    FAI  ANC       261
5745281    CVG  DFW       812
3337632    ATL  BTR       449
510696     ORD  TUS      1437

In [66]:
Top5flightDistance_df = DelayedFlights_df[['origin','dest','distance']].copy()
sorted_flights_df = Top5flightDistance_df.sort_values(by=['distance'], ascending = False, inplace = False).groupby('origin').head(1)
sorted_flights_df = sorted_flights_df.set_axis(['ORIGIN','DESTINATION','DISTANCE'], axis = 1)
sorted_flights_df.set_index('ORIGIN', inplace=True)
sorted_flights_df.head()

DESTINATION  DISTANCE
ORIGIN                      
EWR            HNL      4962
HNL            EWR      4962
ATL            HNL      4502
ORD            HNL      4243
KOA            ORD      4213

Exercici 3 :
Exporta el data set net i amb les noves columnes a Excel.

# Converting back to a CSV

Finally, after extensive work cleaning the data, we are ready to save it as a new CSV file similar to how we read the data with pandas.

In [ ]:
DelayedFlights_df.to_csv('new_DelayedFlights_df.csv')